File: c/src/env_vars.c
#include "env_vars.h"
#include "stdio.h"
#include "stdlib.h"
#include "string.h"

const char* getEnvD(const char* name, const char* defaultValue);
const char* getEnv(const char*);
int getEnvInt(const char*, const char*);

const int Valid = 1;
const int Invalid = 0;
int isVarsValid(EnvVars vars)
{
	if( vars.ApiKey != NULL
		&& vars.Domain != NULL
		&& vars.RecordName != NULL
		&& vars.checkIntervalSeconds > 0
		&& vars.printSameCountStep > 0)
		return Valid;
	return Invalid;
}

EnvVars gEnvVars;

EnvVars* getEnvVars()
{ 
	EnvVars* p = &gEnvVars;
	memset(p, 0, sizeof(gEnvVars));
	p->ApiKey = getEnv("API_KEY");
	p->Domain = getEnv("DOMAIN");
	p->RecordName = getEnv("RECORD_NAME");
	p->IpEcho = getEnvD("IP_ECHO", "http://icanhazip.com");
	p->checkIntervalSeconds = getEnvInt("CHECK_INTERVAL_SECONDS", "60");
	p->printSameCountStep = getEnvInt("PRINT_SAME_COUNT_STEP", "360");

	if(isVarsValid(gEnvVars) == Invalid)
		return NULL;
	
	return p;
}

const char* getEnvD(const char* name, const char* defaultValue)
{
	char* value = getenv(name);
	if(value)
		return value;
	if(defaultValue)
		return defaultValue;

	fprintf(stderr, "please set your [%s] in environment variable\n", name);
	return NULL;
}

const char* getEnv(const char* name)
{
	return getEnvD(name, NULL);
}

int getEnvInt(const char* name, const char* defaultValue)
{
	int i = 1;
	const char *v = getEnvD(name, defaultValue);
	if(v){
		i = atoi(v);
	}
	return i;

}

File: ruby/lib/env_vars.rb

class EnvVars
    attr_accessor :API_KEY,
                    :DOMAIN,
                    :RECORD_NAME,
                    :IP_ECHO,
                    :CHECK_INTERVAL_SECONDS,
                    :PRINT_SAME_COUNT_STEP

    def self.get_var(key, default_value=nil)
        value = ENV[key]
        unless value
            value = default_value
        end

        unless value
            p "Please set your #{key} in environment variable"
        end
        value
    end

    def self.get_var_int(key, default_value=nil)
        value = get_var(key, default_value)
        unless value
            value = default_value
        end
        value.to_i
    end

    def self.Instance
        apiKey = get_var('API_KEY')
        domain = get_var('DOMAIN')
        recordName = get_var('RECORD_NAME')
        ipEcho = get_var('IP_ECHO', "http://icanhazip.com/")
        checkIntervalSeconds = get_var_int('CHECK_INTERVAL_SECONDS', 60)
        printSameCountStep = get_var_int('PRINT_SAME_COUNT_STEP', 6 * 60)

        if apiKey && domain && recordName && ipEcho && checkIntervalSeconds && printSameCountStep
            EnvVars.new(apiKey, domain, recordName, ipEcho, checkIntervalSeconds, printSameCountStep)
        else
            nil
        end
    end

    def initialize(apiKey, domain, recordName, ipEcho, checkIntervalSeconds, printSameCountStep)
        @API_KEY = apiKey
        @DOMAIN = domain
        @RECORD_NAME = recordName
        @IP_ECHO = ipEcho
        @CHECK_INTERVAL_SECONDS = checkIntervalSeconds
        @PRINT_SAME_COUNT_STEP = printSameCountStep
    end
end
